lib/SILOptimizer/Transforms/HoistDestroyAddrs.cpp

```diff
@@ -0,0 +1,448 @@
+/// compiler does not manage the lifetime of raw pointers that escape a lexical
+/// scope. Programmers must manage escaping pointers explicitly via
+/// Builtin.fixLifetime.
+///
+/// Case #3. Synchronization points: If the object potentially has a custom
+/// deinitializer with side effects, then any external function call, which may
+/// contain a memory barrier or system call, prevents hoisting. If the external
+/// function call is annotated as "read-only", then it is safe. Since Swift does
+/// not directly support atomics, no SIL instructions are currently considered
+/// synchronization points.
+///
+///
+/// DestroyAddr hoisting ignores the following deinitialization barriers because
+/// the destroyed memory cannot be an object:
+///
+/// - Aliasing references: Any memory access based on a reference that
+/// potentially aliases the same object.
+///
+/// - Weak reference loads: Any load of a weak or unowned reference that may
+/// alias this object *within the current lexical scope*. The weak reference
+/// need not be accessed. Any use is considered a barrier.
+///
+/// ----------------------------------------------------------------------------
+///
+/// TODO: replace the destroy hoisting in CopyForwarding::forwardCopiesOf and
+/// ensure related tests still pass. However, this requires hoisting over
+/// certain calls. We can do this as long as the call takes a copy of the value
+/// at the root address as an argument, since no deinits will be invoked by the
+/// hoisted destroy.
+///
+/// TODO: enforce an invariant that destroy_addrs jointly post-dominate any
+/// exclusive owned address, that would slightly simplify the algorithm.
+///
+/// ===--------------------------------------------------------------------===//
+
+#include "swift/Basic/GraphNodeWorklist.h"
+#include "swift/SIL/BasicBlockDatastructures.h"
+#include "swift/SIL/MemAccessUtils.h"
+#include "swift/SIL/SILBasicBlock.h"
+#include "swift/SIL/SILBuilder.h"
+#include "swift/SIL/SILInstruction.h"
+#include "swift/SILOptimizer/PassManager/Transforms.h"
+#include "swift/SILOptimizer/Utils/InstructionDeleter.h"
+
+using namespace swift;
+
+static bool isUniqueOwnedAddress(SILValue address) {
+  assert(address->getType().isAddress());
+
+  if (isa<AllocStackInst>(address))
+    return true;
+
+  if (auto *arg = dyn_cast<SILFunctionArgument>(address)) {
+    if (arg->getArgumentConvention() == SILArgumentConvention::Indirect_In)
+      return true;
+  }
+  return false;
+}
+
+// =============================================================================
+// HoistDestroyAddrs Algorithm
+// =============================================================================
+
+class HoistDestroyAddrs {
+  SILValue rootAddress;
+  SILInstruction *addressDefInst = nullptr; // null for function args
+  InstructionDeleter &deleter;
+  bool preserveDebugInfo;
+
+  // MARK: Analysis of the address SSA uses
+
+  bool hasPointerEscape = false;
```

Comment:
Using member variables in this way is as bad as using global variables.
Functions (e.g. `findKnownUses`) should return their results (or return via a reference parameter) and get their inputs as parameters. Otherwise it's a nightmare to figure out the flow of information in the algorithm.

Reply:
@eeckstein I think the code is very clear that these members *are* the result of the SSA-use analysis. `hasPointerEscape` should be declared with the rest of the analysis result, including `addressUsers` and `originalDestroys`.

These are private to this small utility, initialized once by the analysis and subsequently immutable. That's nothing like a global variable!

That said, this utility performs two analysis steps, each with separate results. I usually separate each stage of results into separate classes. In fact, that was also my intention here. The first step, `visitAddressUses`, should be defined in MemAccessUtils.  (This is the only code in this utility with substantial logic, so the utility truly becomes trivial at that point). I didn't do this yet because the existing AccessPath API is massive overkill for this case, and I didn't have time to refactor MemAccessUtils so we have a simpler API for discovering SSA address uses when we don't care about the access path.

Reply:
For sure, those member variables are like global variables within the algorithm. It's not clear what variables are set and used, e.g. by looking at `HoistDestroyAddrs::perform()`.
Why not let the functions in `HoistDestroyAddrs::perform()` return the results and pass inputs as parameters? That would make the flow of information clear.